### DISCUSSION TITLE
How to map inner generic type?

### DISCUSSION THREAD

**Discussion by oraad** [nartc/mapper#499] [Q&A] [answered]
Hi, I'm trying to figure out how to map an inner property of generic type.

Consider the dto class
`class PageableDto<T> {
    number: number
    content: T[]
}`
to be mapped to
`class Pageable<T> {
    page: number
    content: T[]
}`

And I have the possible inner types
`class BookDto {
    title: string;
    author: string;
}`
`class Book {
    title: string;
    author: string;
}`

`class MovieDto {
   title: string;
   director: string;
   releasedOn: Date;
}`
`class Movie {
   title: string;
   director: string;
   date: Date;
}`

How can I map `PageableDto<BookDto> -> Pageable<Book>` and `PageableDto<MovieDto> -> Pageable<Movie>` ?
I'm getting `undefined` for `Pageable`.`content` when I map.

**Comment by jampack** [comment]
bump

**Comment by oraad** [comment, marked as answer]
One solution/workaround I did, is to use `mapDefer`  since I have two types to be passed in this case.

```
createMap(mapper, PageableDto, Pageable,
    forMember(dest => dest.content,
        mapDefer(src => {
            if (src.content?.[0] instanceof BookDto)
                return mapWith(Book, BookDto, src => src.content)

            if (src.content?.[0] instanceof MovieDto)
                return mapWith(Movie, MovieDto, src => src.content)

            // handle all other types not yet specified as is
            return mapFrom(src => src.content)
        })
    )
)
```
The problem with this solution, is that it becomes very verbose if we have more inner types to map from.

**Comment by nartc** [comment]
Hey @oraad, this is a limitation right now with AutoMapper. We will have to think about supporting this with a reasonable API. `mapDefer` would have been my suggestion as well and I am aware of its verbosity.
